cmd/kube-apiserver/app/server.go

```diff
@@ -99,6 +101,11 @@ import (
 const etcdRetryLimit = 60
 const etcdRetryInterval = 1 * time.Second
 
+func init() {
+	// FIXME(anjensan): Find better place / install bulk api into separate registry...
```

Comment:
+1 please fix :)

Reply:
Unfortunately I didn't found a better place for it :)

It is impossible to put such line into apiserver/pkg/apis/bulk/install/install.go due to cycle dependency (legacyscheme lives in /pkg/kubernetes & apiserver can't use packages from).

Do you have any ideas / suggestions?

Reply:
this is a logically separate component, just like the custom resource server and the kube aggregator. those have their own APIs, live in distinct packages (`staging/src/k8s.io/kube-aggregator` and `staging/src/k8s.io/apiextensions-apiserver`), and are combined in https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/server.go#L145

Reply:
I don't think it should be a separate component, it looks more like additional transportation/serialization layer (like HTTP, HTTPS, SPDY...)

Reply:
the distinct API group and expectation that API servers other than kube-apiserver could use  this mechanism to allow bulk operations on their own resources indicates to me it is a separable component. 

it looks like you tried to accomplish that by adding to `k8s.io/apiserver`, but it will be cleaner to follow the pattern of the custom resource server, containing the API and special behaviors into their own component, then install that into the kube-apiserver.

@deads2k can provide guidance here.

Reply:
> Unfortunately I didn't found a better place for it :)

I've fixed original FIXME by moving bulk.{Server,Client}Message objects into a separate registry.
Actually everything was already ready for it :) --
embedded objects (inside WatchEvent/GetResult/ListResult) are still serialized using per-group serializers (taken from genericapiserver.APIGroupInfo).

> the distinct API group

IMO semantically it shouldn't be a separate group. It is closer to v1.WatchEvent.  And it contains 0 resources (both ServerMessage/ClientMessage are not resources).
But it is a good idea to move it into a separate group just because it is highly experimental and it easier to evolve such api (alpha -> beta and so).

>  expectation that API servers other than kube-apiserver could use this mechanism to allow bulk operations 

Ideally external API servers need to do *nothing* in order to enable bulk operations.
Exactly like they don't need to do anything to provide regular 'watch' (http-based).

I've just performed some basic manual testing - it is now possible to watch/list CustomResourceDefinitions through bulk api without any modifications in 'apiextensions-server'.
Are they served from 'apiextensions-server'?

>> containing the API and special behaviors into their own component
The main complication here is that bulk-api should not introduce any "special" behaviors.
It just adds another transport protocol to work with existing apigroups.

Reply:
By definition, it is a new API that does not belong in any other group (because via it you can interact with multiple other groups). That means new endpoints, which don't belong to any existing API group. 

> Ideally external API servers need to do nothing in order to enable bulk operations.

If something else (like the aggregator) is handling the bulk APIs, sure, but this should be structured in a way that a stand-alone server could serve them as well.

> And it contains 0 resources

The top level types submitted to the API are resources.

Reply:
> if something else (like the aggregator) is handling the bulk APIs,

Aggregator already have special handling for bulk. It is part of this PR actually.
Main reason for this -- aggregator should multiplex several websockets (as websocket clients usually can open only 1 connection due to browser policies etc).

> stand-alone server could serve them as well.

Do you mean "a separate server based on genericapiserver"?

> The top level types submitted to the API are resources.

Hm, maybe I was understanding/using the term "resources" a bit incorrectly.
I was though it is something like "entity" which have a separate REST path etc.

For bulk API both bulk.ClientMessage and bulk.ServerMessage are not entities. They should behave exactly like 'GetOptions' / 'WatchEvent'. Maybe I'm registering them incorrectly.